evm/vm/base.py

```diff
@@ -123,29 +122,63 @@ def execute_bytecode(self,
         )
 
     @abstractmethod
-    def make_receipt(self, transaction, computation, state):
+    def make_receipt(self, base_header, transaction, computation, state):
         """
-        Make receipt.
+        Generate the receipt resulting from applying the transaction.
+
+        :param base_header: the header of the block before the transaction was applied.
+        :param transaction: the transaction used to generate the receipt
+        :param computation: the result of running the transaction computation
+        :param state: the resulting state, after executing the computation
+
+        :return: receipt
+        """
+        raise NotImplementedError("Must be implemented by subclasses")
+
+    @abstractmethod
+    def validate_transaction_against_header(self, base_header, transaction):
+        """
+        Validate that the given transaction is valid to apply to the given header.
+
+        :param base_header: header before applying the transaction
+        :param transaction: the transaction to validate
+
+        :raises: ValidationError if the transaction is not valid to apply
         """
         raise NotImplementedError("Must be implemented by subclasses")
 
-    def apply_transaction(self, transaction):
+    def apply_transaction(self, header, transaction):
         """
         Apply the transaction to the vm in the current block.
+
+        :param transaction: to apply
+        :param header: header of the block before application
```

Comment:
Oops, I'll swap the order of these params in the docs. I went back and forth a few times locally. In the end a semi-chronological order made sense to me: the header is in the state before the transaction is applied, so comes first.